Add molar-mass recipe

sergiruiztrepat opened the pull request:
Molar-mass calculates molar mass of a given molecule

### Brief summary of what the package does

Calculates molar mas (or molecular weigth) of a given molecule. 
Ex: M-x molar-mass => H2O => "Molar mass of H2O : 18.015 g/mol (uma)"

### Direct link to the package repository

https://github.com/sergiruiztrepat/molar-mass

### Your association with the package

I am the maintainer.

### Relevant communications with the upstream package maintainer

**None needed**

### Checklist

Please confirm with `x`:

- [X ] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses).
- [ X] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [ X] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [X ] My elisp byte-compiles cleanly
- [ X] `M-x checkdoc` is happy with my docstrings
- [ X] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [ ] I have confirmed some of these without doing them

sergiruiztrepat commented:
Hi! This is my first contribution to melpa and I'm a novice with git and github. I've read contributing.org but I'm not sure if I'm doing things well. I've added only the recipe of my package to the melpa repository. Is it correct?

mattiasb commented:
> Hi! This is my first contribution to melpa and I'm a novice with git and github. I've read contributing.org
> but I'm not sure if I'm doing things well. I've added only the recipe of my package to the melpa repository. Is it correct?

The recipe looks good to me (Note: I'm not a melpa reviewer). @riscy tends to do reviews mostly over the weekends so you might have to wait a bit. :)

mattiasb commented:
Two comments regarding the `elements-mass` variable:
1. You probably want to set it using `defconst`
2. You should rename it to `molar-mass-elements-mass` or something similar to keep the namespacing (this is important to ensure two random packages doesn't clash).

sergiruiztrepat commented:
> Two comments regarding the `elements-mass` variable:
> 
>     1. You probably want to set it using `defconst`
> 
>     2. You should rename it to `molar-mass-elements-mass` or something similar to keep the namespacing (this is important to ensure two random packages doesn't clash).

Thank you! Changes done!

sergiruiztrepat commented:
Hi, @mattiasb , 
I've added an error control function. Maybe you want to review the last commits.
Thanks.

riscy commented:
Thanks both for the above!  Here's a quick first pass --

### molar-mass.el ###

byte-compile (using Emacs 27.1):
```
molar-mass.el:309:53:Warning: `error' called with 1 args to fill 0 format
    field(s)
```

checkdoc (using version 0.6.2):
- No issues!

package-lint (using version 20210127.158):
- No issues!

Other [lints](https://github.com/riscy/melpazoid):
- molar-mass.el#L152: It's safer to sharp-quote function names; use `#'`
- molar-mass.el#L259: It's safer to sharp-quote function names; use `#'`
- It's not necessary to wrap the `else` part of an `if` in a `progn`

A (very quick) closer look:
- Some of the numbers in `molar-mass-elements-mass` have 5 significant digits but others have 3 (e.g. Kr I guess could be expressed with more precision to be consistent with the other elements)
- This should be wrapped in a `let`: `(setq case-fold-search nil)`
- I don't think you need to be wrapping your `(string-match-p` calls with `ignore-errors`
- If you just need truthiness, you can write `(if (string-match-p "[A-Z]" char) t)` as `(string-match-p "[A-Z]" char)`
- It's unusual to lead variable names with dollar signs in elisp (`$cut-list`)
- Why not break `molar-mass-errors` into three functions that could be called by name (thus making the code "self-documenting") rather than inventing error codes that are just used internally?

sergiruiztrepat commented:
Hi @riscy,

Thanks for your review. I will make changes and inform you when it's ready for a new review.

sergiruiztrepat commented:
Hi!

File is ready for review.

There are some issues to comment:

>     * It's not necessary to wrap the `else` part of an `if` in a `progn`

If I delete progn, it works, but byte-compile (Emacs 25.3.2) raises errors with the deleted progn at the
else part:

In molar-mass-pairs-list:
molar-mass.el:195:31:Warning: ‘(setq p2 (concat (cadr elem) (caddr
elem)))’ is a malformed function
molar-mass.el:195:31:Warning: ‘(setq p2 "1")’ is a malformed function

No errors with the progn. Is it because of Emacs version?
 
>     * Some of the numbers in `molar-mass-elements-mass` have 5 significant digits but others have 3 (e.g. Kr I guess could be expressed with more precision to be consistent with the other elements)

I've reviewed all data at `molar-mass-elements-mass` to their maximum precision. I've also added `defcustom molar-mass-significant-digits` to allow the user to customize the significant decimals of the result.
 
>     * I don't think you need to be wrapping your `(string-match-p` calls with `ignore-errors`
> 
>     * If you just need truthiness, you can write `(if (string-match-p "[A-Z]" char) t)` as `(string-match-p "[A-Z]" char)`

I don't know in advance if char is bound or nil. If nil, it will raise an error. I've replaced `ignore-errors` with `(if char`. Is it correct?

riscy commented:
> If I delete progn, it works, but byte-compile (Emacs 25.3.2) raises errors with the deleted progn at the else part

This is because the enclosing parens aren't necessary anymore in this form:
```elisp
       ((and (molar-mass-upcase-p (car elem))
	     (molar-mass-number-p (cadr elem)))
```

Ah I see now.  You could be more succinct with that by writing `(if char (if (string-match-p "[0-9]" char) t))` as `(and char (string-match-p "[0-9]" char))`.  Here's the latest:

```
molar-mass.el:148:1:Warning: defcustom for `molar-mass-significant-digits'
    fails to specify type
molar-mass.el:148:1:Warning: defcustom for `molar-mass-significant-digits'
    fails to specify type
In molar-mass-pairs-list:
molar-mass.el:195:31:Warning: `(setq p2 (concat (cadr elem) (caddr elem)))' is
    a malformed function
molar-mass.el:195:31:Warning: `(setq p2 "1")' is a malformed function
```

sergiruiztrepat commented:
Ok, it is ready for review.
Thanks!

riscy commented:
Thanks!  That puts the checklist well behind us.  I think altogether this looks good, but there are a few more minor issues:

- You're using `print` to output `result-string-format` but it seems like `message` is the correct function to be calling here.  `message` also conveniently takes f-strings so you'd write: ```(message result-string-format data (molar-mass-total-mass (molar-mass-pairs-list elements)))```
- You're using `error` to report errors in the user's input (e.g. `molar-mass-error-in-formula`) but probably you want to use `user-error` which won't accidentally open a big backtrace for the user to try to debug with

Some other minor stuff:
- I know nothing about chemistry but there are 118 known elements and only 114 items in `molar-mass-element-mass` (I didn't want to take too long checking, but about "Oganesson"?)
- It's unusual to lead variable names with dollar signs in elisp ($total-mass)

sergiruiztrepat commented:
>     * I know nothing about chemistry but there are 118 known elements and only 114 items in `molar-mass-element-mass` (I didn't want to take too long checking, but about "Oganesson"?)

Yes, you're right. The four lacking elements can not be used in a compound, but it's ok, I've added them.
Code is ready for review.
Thanks!

riscy commented:
Looks good, and thanks for putting up with my unfamiliarity.  I'll leave one more bit of feedback after experimenting with this -- if you `(require 'thingatpt)` then instead of `read-string` you could use a `completing-read` with a default input value of whatever string is under the cursor:
```elisp
(completing-read "Formula: "
                             (mapcar #'car molar-mass-elements-mass)
                             nil t nil nil
                             (car
                              (assoc
                               (thing-at-point 'word)
                               molar-mass-elements-mass)))
```
Not a blocker but I think that would make this quicker to use in some cases, with better element-discoverability features -- and welcome to MELPA!

sergiruiztrepat commented:
Hi Steve,

Thank you very much. I'm honoured of contributing to MELPA. And I have also learnt a lot in this process.

Two last things:

- I am working in a new utility for chemistry. You will have news soon.

- I am looking for packages to contribute to. I don't want to steal your time, but if you know of somebody who need help, please tell me.

Thanks!
